master/static/srv/ship_logs.py

```diff
@@ -0,0 +1,597 @@
+        assert exit_code is not None
+
+        # Convert signal exits to standard bash-style exits.
+        if exit_code < 0:
+            return 128 - exit_code
+        return exit_code
+
+    finally:
+        # p.kill() is a SIGKILL, so p.wait() should be quick.
+        p.kill()
+        if waiter_started:
+            waiter.join()
+        # After p is dead, the Collectors should run out of input and exit quickly.
+        if stdout_started:
+            stdout.join()
+        if stderr_started:
+            stderr.join()
+        # After everyone else is dead, the shipper could still be in a retry loop for a long time,
+        # so we wait up to DET_LOG_WAIT_TIME seconds for it to exit, then we give up on it.
+        if shipper_started and not shipper_timed_out:
+            shipper.join(timeout=log_wait_time)
+            if shipper.is_alive():
+                # The timeout was reached.
+                logging.error(
+                    f"waited {log_wait_time} seconds for shipper to finish after crash; "
+                    "giving up now"
+                )
+
+
+if __name__ == "__main__":
+    try:
+        logging.basicConfig(
+            format=LOG_FORMAT,
+            stream=sys.stderr,
+        )
+
+        # Emergency debug escape hatch: if the log shipper goes belly-up, you obviously can't expect
+        # to find those logs in task logs, so allow a user to mount a directory into the container
+        # in order to find out what failed.
+        ship_logs_path = os.environ.get("DET_SHIP_LOGS_PATH", "/ship_logs")
+        if os.path.exists(ship_logs_path):
+            try:
+                hostname = os.environ["DET_AGENT_ID"]
+            except Exception:
+                try:
+                    import socket
+
+                    hostname = socket.gethostname()
+                except Exception:
+                    hostname = "unknown"
+            # You can't run logging.basicConfig() twice so we manually add a second handler at the
+            # root logging level.
+            fh = logging.FileHandler(
+                filename=os.path.join(ship_logs_path, f"{hostname}-{time.time()}.log"),
+                # Only create the file if we actually log to it (aka if there's an error).  That way
+                # if there's lots of processes not failing, we don't create tons of empty files.
+                delay=False,
+            )
+            fh.setFormatter(logging.Formatter(LOG_FORMAT))
+            logging.getLogger().addHandler(fh)
+            try:
+                # Touch a single file to indicate that the escape hatch is working, so that in
+                # debugging someone can distinguish "the escape hatch isn't working" from
+                # "ship_logs just isn't hitting any errors".
+                with open(os.path.join(ship_logs_path, "ship-logs-ran"), "w") as f:
+                    pass
+            except Exception:
+                pass
+
+        master_url = os.environ["DET_MASTER"]
```

Comment:
Feels awkward to me, because then we'd have to write a whole commandline interface for the wrapper to specify them to ship_logs.py, when ship_logs.py could just read them from the environment.

It would be a lot more code for literally zero difference in behavior.

Reply:
Thanks for thinking about it. I'm still becoming more comfortable with environment variables as flags. I see what you're saying and agree.